src/com/collibra/server/ServerThread.java

```diff
@@ -36,19 +35,17 @@ public ServerThread(Socket socket, Map<Command, CommandHandler> commandToHandler
     public void run() {
         long startTime = System.nanoTime();
         PrintWriter outData = null;
-        String clientName;
         try {
             BufferedReader inData = new BufferedReader(new InputStreamReader(socket.getInputStream()));
             outData = new PrintWriter(socket.getOutputStream(), true);
             handshake(outData);
             while (true) {
-                ParsedMessage parsedMessage = parseReceivedMessage(readReceivedMessage(inData));
-                clientName = parsedMessage.getClientName();
-                commandToHandler.put(HI, new GreetingCommandHandler(clientName));
-                if (clientName != null) {
-                    commandToHandler.put(BYE, new ByeCommandHandler(clientName, startTime));
+                String receivedMessage = readReceivedMessage(inData);
+                Command command = extractCommand(receivedMessage);
+                if (command == HI) {
+                    commandToHandler.put(BYE, new ByeCommandHandler(extractClientName(receivedMessage), startTime));
```

Comment:
The `BYE` handler should always be present. What if there is no `HI` request sent from the client?

Reply:
It's bug. We have 2 kind of 'Bye'-es i.e. 'BYE Mate!' and 'BYE <name>, WE SPOKE FOR <X> MS... I mixed them. They are different cases.

Reply:
IMO it's not the case as ```BYE``` handler needs the client name. Client name we receive via ```Hi``` request. According to requirements:
````
The server sends a first message back to the client:
[SERVER]HI, I AM <session-id>
where the <session-id> is replaced by the generated session ID. The client will then respond
with:
[CLIENT]HI, I AM <name>
where <name> can be any string of alphanumeric characters + the dash character (-). The server
will respond with:
[SERVER]HI <name>
To end the session, the client will send:
[CLIENT]BYE MATE!
after which the server will respond with:
[SERVER]BYE <name>, WE SPOKE FOR <X> MS
Where <name> is the name the client mentioned in its first message and <X> is the number of
milliseconds the session.
Note: this message should also be sent to the client when the session times out (after not
receiving any message from the client for 30 seconds).
````